flang/lib/Lower/IO.cpp

```diff
@@ -113,29 +115,29 @@ static mlir::FuncOp getIORuntimeFunc(Fortran::lower::FirOpBuilder &builder) {
   return func;
 }
 
-/// Generate a call to end an IO statement
+/// Generate calls to end an IO statement.  Return the IOSTAT value, if any.
+/// It is the caller's responsibility to generate branches on that value.
 static mlir::Value genEndIO(Fortran::lower::AbstractConverter &converter,
                             mlir::Location loc, mlir::Value cookie,
-                            const ErrorHandling &eh) {
-  // Terminate IO
+                            const ConditionSpecifierInfo &csi) {
   auto &builder = converter.getFirOpBuilder();
-  if (eh.ioMsg) {
-    TODO(); // eh.ioMsgExpr is never set
-    auto getMsg = getIORuntimeFunc<mkIOKey(GetIoMsg)>(builder);
-    auto var = converter.genExprAddr(eh.ioMsgExpr, loc);
-    mlir::Value varLen{}; // FIXME
-    llvm::SmallVector<mlir::Value, 3> args{cookie, var, varLen};
-    builder.create<mlir::CallOp>(loc, getMsg, args);
+  if (csi.ioMsgExpr) {
+    auto getIoMsg = getIORuntimeFunc<mkIOKey(GetIoMsg)>(builder);
+    auto ioMsgVar = converter.genExprAddr(csi.ioMsgExpr, loc);
+    auto ioMsgVarLen = builder.createUnboxChar(ioMsgVar).second;
+    llvm::SmallVector<mlir::Value, 3> args{cookie, ioMsgVar, ioMsgVarLen};
+    builder.create<mlir::CallOp>(loc, getIoMsg, args);
   }
-  auto endIOFunc = getIORuntimeFunc<mkIOKey(EndIoStatement)>(builder);
+  auto endIoStatement = getIORuntimeFunc<mkIOKey(EndIoStatement)>(builder);
```

Comment:
`EndIoStatement()` is a function that signals the end of an I/O statement and not a statement.

Reply:
The idea is to use the actual function name when building a single known call.  Call creation then reads nicely:
`auto call = builder.create<mlir::CallOp>(loc, endIoStatement, endArgs);`

Reply:
Ok. I just was wondering if this might be confusing to someone new.